Asterix/optics/propagation_functions.py

```diff
@@ -416,3 +414,136 @@ def fft_choosecenter(image, inverse=False, center_pos='bb', norm='backward'):
     farray *= np.exp(sens * (2. * 1j * np.pi / np.sqrt(Nx * Ny)) * np.sum(direct * fourier))
 
     return farray
+
+
+def butterworth_circle(dim, size_filter, order=5, xshift=0, yshift=0):
+    """
+    Return a circular Butterworth filter.
+
+    AUTHOR: Raphaël Galicher (in IDL)
+            ILa (to Python)
+
+    Parameters
+    ----------
+    dim : int
+        Dimension of 2D output array in pixels. If even, filter will be centered on a pixel, but can be shifted to
+        between pixels by using xshift=-0.5 and yshift=-0.5. If uneven, filter will be centered between pixels.
+    size_filter : int
+        Inverse size of the filter.
+    order : int
+        Order of the filter.
+    xshift : float
+        Shift of filter with respect to its array in the x direction, in pixels.
+    yshift : float
+        Shift of filter with respect to its array in the y direction, in pixels.
+
+    Returns
+    -------
+    2D array
+
+    Example
+    --------
+    siz = 100
+    rad = int(siz / 2)
+
+    bfilter1 = butterworth_circle(siz, rad, order=1, xshift=-0.5, yshift=-0.5)
+    bfilter2 = butterworth_circle(siz, rad, order=2, xshift=-0.5, yshift=-0.5)
+    bfilter3 = butterworth_circle(siz, rad, order=3, xshift=-0.5, yshift=-0.5)
+    bfilter4 = butterworth_circle(siz, rad, order=4, xshift=-0.5, yshift=-0.5)
+    bfilter5 = butterworth_circle(siz, rad, order=5, xshift=-0.5, yshift=-0.5)
+    bfilter6 = butterworth_circle(siz, rad, order=6, xshift=-0.5, yshift=-0.5)
+
+    plt.figure(figsize=(16,8))
+    plt.subplot(1,2,1)
+    plt.imshow(bfilter5, cmap='Greys_r', origin='lower')
+    plt.title("Order = 5")
+    plt.colorbar()
+    plt.subplot(1,2,2)
+    plt.plot(bfilter1[rad], label="order=1")
+    plt.plot(bfilter2[rad], label="order=2")
+    plt.plot(bfilter3[rad], label="order=3")
+    plt.plot(bfilter4[rad], label="order=4")
+    plt.plot(bfilter5[rad], label="order=5")
+    plt.plot(bfilter6[rad], label="order=6")
+    plt.legend()
+    plt.show()
+    """
+    ty = (np.arange(dim) - yshift - dim / 2)
+    tx = (np.arange(dim) - xshift - dim / 2)
+    xx, yy = np.meshgrid(ty, tx)
+
+    butterworth = 1 / np.sqrt(1 + (np.sqrt(xx ** 2 + yy ** 2) / np.abs(size_filter) * 2) ** (2. * order))
+
+    return butterworth
+
+
+def prop_fpm_regional_sampling(pup, fpm, nbres=np.array([0.1, 5, 50, 100]), samp_outer=2, filter_order=15, alpha=1.5):
```

Comment:
Can you pass if dir_save_all_planes as parameter and allow a full save if dir_save_all_planes is not None like for other propagation method. it's always useful for debugging.

Reply:
This is handled in the `EF_through()` method of the class `Coronagraph`, like for all the other propagation methods.

Reply:
Or do you mean to also access the focal plane?

Reply:
yeah, look at the way I'm doing the mft propagation in the coronagraph

Reply:
Yes, but the regional propagation goes straight from the "apodizer" plane to the Lyot plane. If I wanted to have access to the focal plane of the FPM, I would have to pass these parameters to the function handling the back-and-forth propagations which is not how the other propagation methods do it. Would you still like me to implement this?